cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -121,6 +123,108 @@ SQLRETURN SQLFreeHandle(SQLSMALLINT type, SQLHANDLE handle) {
   return SQL_ERROR;
 }
 
+SQLRETURN SQLGetDiagFieldW(SQLSMALLINT handleType, SQLHANDLE handle,
+                           SQLSMALLINT recNumber, SQLSMALLINT diagIdentifier,
+                           SQLPOINTER diagInfoPtr, SQLSMALLINT bufferLength,
+                           SQLSMALLINT* stringLengthPtr) {
+  using driver::odbcabstraction::Diagnostics;
+  using ODBC::GetStringAttribute;
+  using ODBC::ODBCConnection;
+  using ODBC::ODBCEnvironment;
+
+  if (!handle) {
+    return SQL_INVALID_HANDLE;
+  }
+
+  if (!diagInfoPtr) {
```

Comment:
It's legal to pass a NULL here if they just want the length. Same as for attributes.